alton/pause_event.py

```diff
@@ -0,0 +1,373 @@
+        Construct the file name of a YAML file holding pause event data.
+        """
+        return '{event_id}_{time_str}_{pipeline_system}.yml'.format(
+            event_id=event_id, time_str=time_str, pipeline_system=pipeline_system
+        )
+
+    def add_pipeline_event(self, who_paused, pipeline_system, pause_reason):
+        """
+        Pauses a pipeline system from being released.
+
+        Arguments:
+            who_paused (str): HipChat username which paused the pipeline.
+            pipeline_system (str): Pipeline system name to pause, e.g. edxapp, ecommerce, etc.
+            pause_reason (str): Reason why pipeline system was paused.
+
+        Returns:
+            status (dict): Dictionary containing the keys:
+                event_id (str): Event ID of added pause event.
+                paused (bool): True if pipeline system was paused due to pause event addition.
+
+        Raises:
+        """
+        # Record the number of pause events before adding a new one.
+        num_initial_pause_events = len(self._get_current_pause_events(pipeline_system))
+
+        # Capture the current date/time as a string.
+        current_time = datetime.now()
+        current_time_str = current_time.strftime(self.TIME_FORMAT)
+        # Hash the date/time to create a unique pause event ID - use the datetime obj for msec-uniqueness.
+        event_id = hashlib.sha1(unicode(current_time)).hexdigest()[-8:]
+        event_contents = {
+            'event_id': event_id,
+            'pipeline_system': pipeline_system,
+            'who_paused': who_paused,
+            'time_paused': current_time_str,
+            'who_cleared': None,
+            'time_cleared': None,
+            'pause_reason': pause_reason,
+        }
+        # Create the current pause file.
+        pause_file_name = self._make_pause_event_filename(event_id, current_time_str, pipeline_system)
+        current_pause_filepath = '{current_dir}{pause_file}'.format(
+            current_dir=self.CURRENT_DIRECTORY,
+            pause_file=pause_file_name
+        )
+        self._create_s3_file(current_pause_filepath, yaml.safe_dump(event_contents))
+
+        # Create the historical pause file.
+        history_pause_filepath = self._make_history_pause_filepath(current_time, pause_file_name)
+        self._create_s3_file(history_pause_filepath, yaml.safe_dump(event_contents))
+
+        # If first pause event for the pipeline system, pause the GoCD pipelines.
+        if num_initial_pause_events == 0:
+            log.info("First pause event for pipeline system '%s' - pausing.", pipeline_system)
+            # Pause the pipeline system.
+            for pipeline_name in PIPELINE_SYSTEM_INFO[pipeline_system]:
+                log.info("Pausing pipeline '%s'.", pipeline_name)
+                self.gocd_client.pause_pipeline(pipeline_name, pause_reason)
+
+        pause_status = {
+            'event_id': event_id,
+            'paused': num_initial_pause_events == 0
```

Comment:
Won't this be incorrect on the first pause event? Because if it's the first pause event `num_initial_pause_events` will be 0 but at this point in the code the number of pause events would actually be 1 so the pipeline would be paused.